### ISSUE TITLE
add methods

### ISSUE THREAD

**Issue by Andron00e** [epfml/llm-baselines#18]
* ~~SOAP~~
* Muon
* Shampoo (basic)
* Adam-mini
* LION
*  ...

**Comment by Andron00e** [comment]
add schedules also | [link](https://github.com/epfml/schedules-and-scaling/blob/main/src/main.py)

**Comment by Andron00e** [comment]
some problems with installation of the lates version ```schedulefree```, so I added this manually
see: https://github.com/epfml/llm-baselines/blob/soap/src/optim/schedulefree.py

**Comment by martinjaggi** [comment]
is there a pull request for this? would be nice to collaborate

**Comment by Andron00e** [comment]
> is there a pull request for this? would be nice to collaborate

hi, we are deploying it to the ```soap``` branch together with @mpagli

**Comment by Andron00e** [comment]
a useful settings:
* [anything but sgd](https://arxiv.org/abs/2407.07972)
* ...

**Comment by Andron00e** [comment]
**Adam-mini Note**

I use ```model.named_parameters()``` for Adam-mini instead of ```group_specs```, therefore in the ```main.py``` it looks like:

```
  elif args.opt == "adam-mini":
      opt = Adam_mini(
          device=args.device,
          world_size=args.world_size,
          named_parameters=model.named_parameters(),  # check
          lr=args.lr,
          betas=(args.beta1, args.beta2),
          weight_decay=args.weight_decay,
          model_sharding=args.model_sharding,
          dim=args.n_embd,
          n_heads=args.n_head,
          n_kv_heads=args.n_kv_head,
          verbose=args.adam_mini_verbose,
      )
```

**Comment by kylematoba** [comment]
hi, I'll add sophia and adafactor.

**Comment by Andron00e** [comment]
> hi, I'll add sophia and adafactor.

Hello! Super, just develop this in your branch and then PR to `soap`. I am a bit overloaded these days, but wanted to try Sophia also

Note: in official repository, they do not show SophiaH (with Hutchinson's preconditioner), only SophiaG. We want to have both methods here. SophiaH is nicely implemented in optax for now, but its not so hard to write in PyToch, see: [this link](https://github.com/stanford-crfm/levanter/blob/331c0aa02eec635fa220fc44267cede455b1bca2/src/levanter/optim/sophia.py)

Thx)

**Comment by martinjaggi** [comment]
muon optimizer should also be a good one to add. i think @doikov might be interested in that one too:
https://x.com/Yuchenj_UW/status/1846964136204173318

**Comment by martinjaggi** [comment]
once we have a handful, we'll have a nice benchmark collection for LLM optimizers, probably worth a small writeup soon

**Comment by Andron00e** [comment]
> muon optimizer should also be a good one to add. i think @doikov might be interested in that one too: https://x.com/Yuchenj_UW/status/1846964136204173318

yes, i am working on that, just busy with iclr reviewing these days. 
already have some test runs of the Muon. but, again, it is hard to deduce when ```batch size``` is less than 0.5M tokens

**Comment by Andron00e** [comment]
> hi, Bristen is back early, so I'll get back to that.
> 
> I did some research on Sophia, though, main findings:
> 
> * The official implementation of SophiaG makes some weird choices, described here https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1244/final-projects/CaiaMaiCostelloJasonDanielLazar.pdf.
> * The levanter implementation does not have SophiaG, only SophiaH.
> * There's a quite readable Julia implementation of SophiaH here https://github.com/SciML/Optimization.jl/blob/master/src/sophia.jl.
> 
> Adafactor is simple, it's already close to being released officially, see [pytorch/pytorch#129905](https://github.com/pytorch/pytorch/pull/129905).
> 
> When I get some time next I'll return to this if you haven't.

I mean, for the official version of SophiG, you may just look at the paper's repo: https://github.com/Liuhong99/Sophia

**Comment by Andron00e** [comment]
> hi, Bristen is back early, so I'll get back to that.
> 
> I did some research on Sophia, though, main findings:
> 
> * The official implementation of SophiaG makes some weird choices, described here https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1244/final-projects/CaiaMaiCostelloJasonDanielLazar.pdf.
> * The levanter implementation does not have SophiaG, only SophiaH.
> * There's a quite readable Julia implementation of SophiaH here https://github.com/SciML/Optimization.jl/blob/master/src/sophia.jl.
> 
> Adafactor is simple, it's already close to being released officially, see [pytorch/pytorch#129905](https://github.com/pytorch/pytorch/pull/129905).
> 
> When I get some time next I'll return to this if you haven't.

I mean, for the official version of SophiG, you may just look at the paper's repo: https://github.com/Liuhong99/Sophia
